device-registry/src/main/scala/org/genivi/sota/device_registry/PublicCredentialsResource.scala

```diff
@@ -0,0 +1,88 @@
+/**
+  * Copyright: Copyright (C) 2016, ATS Advanced Telematic Systems GmbH
+  * License: MPL-2.0
+  */
+package org.genivi.sota.device_registry
+
+import akka.Done
+import akka.http.scaladsl.marshalling.Marshaller._
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server.{Directive1, Route}
+import akka.stream.ActorMaterializer
+import akka.stream.scaladsl.{Sink, Source}
+import akka.util.ByteString
+import org.genivi.sota.data.{Device, DeviceT, Namespace, Uuid}
+import org.genivi.sota.device_registry.db.{PublicCredentialsRepository, DeviceRepository}
+import org.genivi.sota.http.{AuthedNamespaceScope, Scopes}
+import org.genivi.sota.marshalling.CirceMarshallingSupport._
+import org.genivi.sota.messaging.MessageBusPublisher
+import org.genivi.sota.messaging.Messages.{DeviceCreated, DevicePublicCredentialsSet}
+import org.slf4j.LoggerFactory
+import slick.driver.MySQLDriver.api._
+import scala.concurrent.{ExecutionContext, Future}
+
+class PublicCredentialsResource(authNamespace: Directive1[AuthedNamespaceScope],
+                          messageBus: MessageBusPublisher,
+                          deviceNamespaceAuthorizer: Directive1[Uuid])
+                         (implicit db: Database,
+                          mat: ActorMaterializer,
+                          ec: ExecutionContext) {
+  import Device._
+
+  def fetchPublicCredentials(uuid: Uuid): Route = {
+    complete(db.run(PublicCredentialsRepository.findByUuid(uuid)))
+  }
+
+  private[this ]val log = LoggerFactory.getLogger(this.getClass)
+  // FIXME: There is a bug in akka, so we need to drain the stream before
+  // returning the response
+  private def drainStream(file: Source[ByteString, Any]): Future[Done] = {
+    file.runWith(Sink.ignore).recoverWith { case t =>
+      log.info(s"Could not drain stream: ${t.getMessage}")
+      Future.successful(Done)
+    }
+  }
+
+  def createDeviceWithPublicCredentials(ns: Namespace): Route = {
+    // parameter because formFields doesen't work
+    // "Substream Source cannot be materialized more than once"
+    parameter('deviceId) { pdevId =>
```

Comment:
Hmm that is true that I could change `DeviceT`, that would make this simpler and not having to work with this multipart stuff.